### PR TITLE
Run unit tests on more backends

### PR DESCRIPTION
**Pull request by nirbheek**
Initial commits lay the groundwork for this, but don't run it on anything except Ninja yet.

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -350,7 +334,11 @@ def setUp(self):
         self.mconf_command = [sys.executable, os.path.join(src_root, 'mesonconf.py')]
         self.mintro_command = [sys.executable, os.path.join(src_root, 'mesonintrospect.py')]
         self.mtest_command = [sys.executable, os.path.join(src_root, 'mesontest.py'), '-C', self.builddir]
-        self.ninja_command = [detect_ninja(), '-C', self.builddir]
+        # Backend-specific commands
+        self.backend = os.environ.get('MESON_BACKEND', 'ninja')
+        self.build_command, self.clean_command, self.test_command, self.install_command, \
+                self.uninstall_command = get_backend_commands(self.backend)
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: run_project_tests.py [hunk 2]
```diff
@@ -155,39 +156,22 @@ def stop_handler(signal, frame):
...
-        if do_debug:
-            compile_commands = [ninja_command, '-v']
-        else:
-            compile_commands = [ninja_command]
-        compile_commands += ['-w', 'dupbuild=err']
-        test_commands = [ninja_command, 'test', 'benchmark']
-        install_commands = [ninja_command, 'install']
-        clean_commands = [ninja_command, 'clean']
+    compile_commands, clean_commands, test_commands, install_commands, \
+            uninstall_commands = get_backend_commands(backend, do_debug)
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: run_tests.py [hunk 3]
```diff
@@ -20,15 +20,84 @@
...
+    exe_suffix = '.exe'
+else:
+    exe_suffix = ''
+
+def get_backend_args_for_dir(backend, builddir):
+    '''
+    Visual Studio backend needs to be given the solution to build
+    '''
+    if backend.startswith('vs'):
+        sln_name = glob(os.path.join(builddir, '*.sln'))[0]
```
Comment (ignatenkobrain) [inline run_tests.py]: __[flake8]__
* [F821] undefined name 'glob'

**Comment by nirbheek** [comment]
Unit tests are failing right now with `msbuild` because I'm not entirely sure what the syntax is to build a specific target in a specific project. Will update it when I can.

File: run_project_tests.py [hunk 4]
```diff
@@ -152,50 +154,36 @@ def stop_handler(signal, frame):
...
+            backend = 'xcode'
+        else:
+            backend = 'ninja'
+    # Set backend arguments for Meson
+    if backend.startswith('vs'):
         backend_flags = ['--backend=' + backend]
-        compile_commands = ['msbuild']
-        test_commands = ['msbuild', 'RUN_TESTS.vcxproj']
-    elif backend == 'xcode' or (backend is None and mesonlib.is_osx()):
+        backend = Backend.vs
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 5]
```diff
@@ -152,50 +154,36 @@ def stop_handler(signal, frame):
...
-        compile_commands += ['-w', 'dupbuild=err']
-        test_commands = [ninja_command, 'test', 'benchmark']
-        install_commands = [ninja_command, 'install']
-        clean_commands = [ninja_command, 'clean']
-
-def get_compile_commands_for_dir(compile_commands, test_build_dir):
-    if 'msbuild' in compile_commands[0]:
-        sln_name = glob(os.path.join(test_build_dir, '*.sln'))[0]
-        comp = compile_commands + [os.path.split(sln_name)[-1]]
+        backend = Backend.xcode
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 6]
```diff
@@ -152,50 +154,36 @@ def stop_handler(signal, frame):
...
+    elif backend == 'ninja':
+        backend_flags = ['--backend=ninja']
+        backend = Backend.ninja
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 7]
```diff
@@ -440,15 +414,15 @@ def detect_tests_to_run():
...
-    all_tests.append(('vala', gather_tests('test cases/vala'), False if using_backend('ninja') and shutil.which('valac') else True))
-    all_tests.append(('rust', gather_tests('test cases/rust'), False if using_backend('ninja') and shutil.which('rustc') else True))
-    all_tests.append(('d', gather_tests('test cases/d'), False if using_backend('ninja') and have_d_compiler() else True))
-    all_tests.append(('objective c', gather_tests('test cases/objc'), False if using_backend(('ninja', 'xcode')) and not mesonlib.is_windows() else True))
-    all_tests.append(('fortran', gather_tests('test cases/fortran'), False if using_backend('ninja') and shutil.which('gfortran') else True))
-    all_tests.append(('swift', gather_tests('test cases/swift'), False if using_backend(('ninja', 'xcode')) and shutil.which('swiftc') else True))
-    all_tests.append(('python3', gather_tests('test cases/python3'), False if using_backend('ninja') and shutil.which('python3') else True))
+    all_tests.append(('java', gather_tests('test cases/java'), False if backend is Backend.ninja and not mesonlib.is_osx() and have_java() else True))
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 8]
```diff
@@ -440,15 +414,15 @@ def detect_tests_to_run():
...
+    all_tests.append(('C#', gather_tests('test cases/csharp'), False if backend is Backend.ninja and shutil.which('mcs') else True))
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 9]
```diff
@@ -440,15 +414,15 @@ def detect_tests_to_run():
...
+    all_tests.append(('vala', gather_tests('test cases/vala'), False if backend is Backend.ninja and shutil.which('valac') else True))
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 10]
```diff
@@ -440,15 +414,15 @@ def detect_tests_to_run():
...
+    all_tests.append(('rust', gather_tests('test cases/rust'), False if backend is Backend.ninja and shutil.which('rustc') else True))
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 11]
```diff
@@ -440,15 +414,15 @@ def detect_tests_to_run():
...
+    all_tests.append(('d', gather_tests('test cases/d'), False if backend is Backend.ninja and have_d_compiler() else True))
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 12]
```diff
@@ -440,15 +414,15 @@ def detect_tests_to_run():
...
+    all_tests.append(('objective c', gather_tests('test cases/objc'), False if backend in (Backend.ninja, Backend.xcode) and not mesonlib.is_windows() else True))
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 13]
```diff
@@ -440,15 +414,15 @@ def detect_tests_to_run():
...
+    all_tests.append(('fortran', gather_tests('test cases/fortran'), False if backend is Backend.ninja and shutil.which('gfortran') else True))
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 14]
```diff
@@ -440,15 +414,15 @@ def detect_tests_to_run():
...
+    all_tests.append(('swift', gather_tests('test cases/swift'), False if backend in (Backend.ninja, Backend.xcode) and shutil.which('swiftc') else True))
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

File: run_project_tests.py [hunk 15]
```diff
@@ -440,15 +414,15 @@ def detect_tests_to_run():
...
+    all_tests.append(('python3', gather_tests('test cases/python3'), False if backend is Backend.ninja and shutil.which('python3') else True))
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'Backend'

**Comment by nirbheek** [comment]
PR should be good to go!

**Comment by jpakkane** [comment]
But it conflicts. :(

**Pull request merged**